File: java_solutions/1419.minimum-number-of-frogs-croaking.java
/*
 * @lc app=leetcode id=1419 lang=java
 *
 * [1419] Minimum Number of Frogs Croaking
 */

class Solution {
    public int minNumberOfFrogs(String croakOfFrogs) {
        return java.util.Optional.of(new int[][]{new int[5], {0}})
            .filter(s -> croakOfFrogs.chars()
                .map("croak"::indexOf)
                .allMatch(i -> i != -1 
                    && (s[0][i] += 1) > 0 
                    && (i == 0 || s[0][i] <= s[0][i - 1]) 
                    && (s[1][0] = java.lang.Math.max(s[1][0], s[0][0] - s[0][4])) >= 0))
            .filter(s -> java.util.Arrays.stream(s[0]).allMatch(v -> v == s[0][0]))
            .map(s -> s[1][0])
            .orElse(-1);
    }
}

File: kotlin_solutions_1/1419.minimum-number-of-frogs-croaking.kt
/*
 * @lc app=leetcode id=1419 lang=kotlin
 *
 * [1419] Minimum Number of Frogs Croaking
 */

class Solution { fun minNumberOfFrogs(croakOfFrogs: String): Int = croakOfFrogs.fold(Triple(IntArray(5), 0, true)) { (c, m, v), ch -> "croak".indexOf(ch).let { i -> if (i == -1 || !v) Triple(c, m, false) else c.also { it[i]++ }.let { if (i > 0 && c[i] > c[i - 1]) Triple(c, m, false) else Triple(c, maxOf(m, c[0] - c[4]), true) } } }.let { (c, m, v) -> if (v && c.all { it == c[0] }) m else -1 } }